kreuzberg/_ocr/_table_extractor.py

```diff
@@ -0,0 +1,402 @@
+
+import numpy as np
+
+from kreuzberg.exceptions import ParsingError
+
+if TYPE_CHECKING:
+    from kreuzberg._types import TSVWord
+
+try:
+    from scipy.cluster.hierarchy import fclusterdata
+
+    SCIPY_AVAILABLE = True
+except ImportError:
+    SCIPY_AVAILABLE = False
+
+
+@runtime_checkable
+class TableExtractor(Protocol):
+    """Protocol for table extraction algorithms."""
+
+    def extract_words(self, tsv_data: str) -> list[TSVWord]:
+        """Parse TSV data into word objects.
+
+        Args:
+            tsv_data: Raw TSV output from Tesseract.
+
+        Returns:
+            List of word dictionaries with position and text data.
+        """
+        ...
+
+    def detect_columns(self, words: list[TSVWord]) -> list[int]:
+        """Detect column positions using clustering.
+
+        Args:
+            words: List of word dictionaries from TSV.
+
+        Returns:
+            Sorted list of column X positions.
+        """
+        ...
+
+    def detect_rows(self, words: list[TSVWord]) -> list[int]:
+        """Detect row positions using clustering.
+
+        Args:
+            words: List of word dictionaries from TSV.
+
+        Returns:
+            Sorted list of row Y positions.
+        """
+        ...
+
+    def reconstruct_table(self, words: list[TSVWord]) -> list[list[str]]:
+        """Reconstruct table as 2D array.
+
+        Args:
+            words: List of word dictionaries from TSV.
+
+        Returns:
+            2D list representing the table structure.
+        """
+        ...
+
+    def to_markdown(self, table: list[list[str]]) -> str:
+        """Convert table to markdown format.
+
+        Args:
+            table: 2D list representing the table.
+
+        Returns:
+            Markdown-formatted table string.
+        """
+        ...
+
+
+@dataclass
+class TesseractTableExtractor:
+    """Concrete implementation of table extraction from TSV.
+
+    This class implements algorithms for detecting table structure
+    from Tesseract's TSV output and reconstructing it as formatted tables.
+    """
+
+    column_threshold: int = 20
+    """Pixel threshold for column clustering."""
+
+    row_threshold_ratio: float = 0.5
+    """Row threshold as ratio of mean text height."""
+
+    min_confidence: float = 30.0
+    """Minimum confidence score to include a word."""
+
+    def extract_words(self, tsv_data: str) -> list[TSVWord]:
+        """Parse TSV output into structured word data.
+
+        Args:
+            tsv_data: Raw TSV output from Tesseract.
+
+        Returns:
+            List of word dictionaries with position and text data.
+
+        Raises:
+            ParsingError: If TSV data cannot be parsed.
+        """
+        try:
+            reader = csv.DictReader(StringIO(tsv_data), delimiter="\t")
+            words: list[TSVWord] = []
+
+            for row in reader:
+                if row.get("level") == "5" and row.get("text", "").strip():
```

Comment:
this is highly nested and badly written logic, try, for, if, try, if, all heavily nested, this entire file requires a refactor and cleanup pass